Update plugin embedded docs (#4568)
SonarSource/sonar-dotnet: sonar-csharp-plugin/src/main/resources/documentation.md

To analyze tool-generated code (e.g. WCF code generated by `SvcUtil.exe`, protobuf code generated by `protoc`, Swagger client code generated by `NSwag`) for a specific C# project, enable the "Analyze generated code" setting inside **Project > Administration > General Settings > C#**. By default, tool-generated code files are skipped from analysis.

> Generated code detection is based on file name, special comments and attributes. Currently recognized values are in [GeneratedCodeRecognizer.cs](https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/src/SonarAnalyzer.Common/Helpers/GeneratedCodeRecognizer.cs).

Reviewer: ```suggestion
The detection of generated code is based on the file name, special comments, and attributes. The currently recognized values are in [GeneratedCodeRecognizer.cs](https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/src/SonarAnalyzer.Common/Helpers/GeneratedCodeRecognizer.cs).
```

Reviewer: (BTW I installed the Grammarly browser extension which helps with hints)

## Scanner compatibility

To analyze C# code, you need to use the [SonarScanner for .NET](/analysis/scan/sonarscanner-for-msbuild/) version 4.x or newer.

## Exclusions

> Files excluded from analysis should be set in a project configuration. This exclusion still performs the analysis on the file and filters final results.

Reviewer: ```suggestion
The files which are to be excluded from analysis should be set in the project configuration. The files are still going to be analyzed during the MSBuild compilation, the exclusion happening as a filter before sending the results to the server.
```

Author: Are you sure about the `which are to be`? That one sounds very complicated/future-conditional-ish.

Author: `the exclusion happening as a` - isn't there a verb missing? `the exclusion is happening as a`

Author: The `before sending the results to the server` is providing too much technical details IMO.

Reviewer: the idea is that 
> This exclusion still performs the analysis on the file and filters final results.
Is not clear at all, because:
- we say "the exclusion performs the analysis", which isn't the case (the exclusion doesn't perform anything)
- either way, our users don't know when the analysis is done, that it is done during the build etc

AND I agree with you it's too much detail :). So a new version:

> The files which are to be excluded from analysis should be set in the project configuration. The files are still going to be analyzed during the MSBuild compilation. The configuration of exclusions is used to filter the analysis results.

> As an alternative, `.editorconfig` file can be used to disable analysis for a specific rule on a file or directory. This can solve performance problems on large files.

Reviewer: ```suggestion
As an alternative, an `.editorconfig` file can be used to disable the analysis for a specific rule on a file or directory. This can solve performance problems on large files. 
```

```
[Path/File.cs]

> dotnet_diagnostic.Sxxx.severity = none

Reviewer: We have a problem that the class name appears in `reportanalyzer` logs and the ID is hard to get... Maybe we should mention the version of Roslyn where this works, and also how to get the rule ID

Author: I'd keep that as a community guide about performance troubleshooting. This is analyzer docs and /p:ReportAnalyzer is compiler stuff.

```

> 

Reviewer: ```suggestion

You can read more in the [official documentation](https://docs.microsoft.com/en-us/visualstudio/code-quality/use-roslyn-analyzers?view=vs-2019#manually-configure-rule-severity-in-an-editorconfig-file).
```

Author: I intentionally avoided that, as it's VS/dotnet stuff that's not specific to our analyzer. And docs show a very easy way how to disable rules contrary your QP/QG settings.

To analyze tool-generated code (e.g. WCF code generated by `SvcUtil.exe`, protobuf code generated by `protoc`, Swagger client code generated by `NSwag`) for a specific C# project, enable the "Analyze generated code" setting inside **Project > Administration > General Settings > C#**. By default, tool-generated code files are skipped from analysis.

> The detection of generated code is based on file name, special comments and attributes. The currently recognized values are in [GeneratedCodeRecognizer.cs](https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/src/SonarAnalyzer.Common/Helpers/GeneratedCodeRecognizer.cs).

Reviewer: ```suggestion
The detection of generated code is based on file name, special comments, and attributes.
```

> The detection of generated code is based on file name, special comments and attributes. The currently recognized values are in [GeneratedCodeRecognizer.cs](https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/src/SonarAnalyzer.Common/Helpers/GeneratedCodeRecognizer.cs).

Reviewer: ```suggestion
The detection of generated code is based on file name, special comments, and attributes. The currently recognized values are in [GeneratedCodeRecognizer.cs](https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/src/SonarAnalyzer.Common/Helpers/GeneratedCodeRecognizer.cs).
```

Reviewer: grammarly says there a comma needed there. @tom-howlett-sonarsource could you confirm/infirm?

Reviewer 2: I believe Grammarly is proposing an [Oxford Comma](https://www.grammarly.com/blog/what-is-the-oxford-comma-and-why-do-people-care-so-much-about-it/) here. I personally wouldn't use one here, feels unnecessary.

## Scanner compatibility

To analyze C# code, you need to use the [SonarScanner for .NET](/analysis/scan/sonarscanner-for-msbuild/) version 4.x or newer.

## Exclusions

> Files excluded from analysis should be set in the project configuration. The files are still going to be analyzed during the compilation and results will be filtered according to the exclusion settings.

Reviewer: ```suggestion
The files which are to be excluded from analysis should be set in the project configuration. Excluded files are still going to be analyzed during the compilation and the results will be filtered according to the exclusion settings.
```

Reviewer: (you can also ask @tom-howlett-sonarsource as he's a native English speaker)

Author: Files to be exluded....

Reviewer 2: Either read fine. The second starts to feel a bit wordy - "Files **to be** excluded should..." works for me
